agent/csi/volumes.go

```diff
@@ -0,0 +1,259 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu        sync.RWMutex                     // To sync between Add(), Remove() and Reset() for "m" map
+	tryMu     sync.RWMutex                     // To sync between tryAddVolume() and tryRemoveVolume()
+	mapMu     sync.RWMutex                     // To sync access to "pluginMap" map
```

Comment:
There are a lot of locks here, and several methods acquire more than one at the same time. Take extra care to ensure there can't be any deadlocks, or simply and reduce the number of locks used.

Reply:
Reduced to two